# Phase 3: Data Understanding

## Objective

The purpose of this phase is to understand the structure, content, quality, and meaning of the Walmart sales dataset before performing any cleaning or transformation.

This phase will examine:

- Dataset dimensions
- Column names
- Data types
- First and last records
- Missing values
- Duplicate records
- Unique values
- Summary statistics
- Date range
- Target variable
- Feature types

## Why Data Understanding Is Important

Before cleaning or modelling a dataset, a data scientist must understand what the data contains.

This helps to identify:

- Incorrect data types
- Missing values
- Duplicate records
- Unexpected categories
- Possible data-quality problems
- The target variable
- Suitable features for machine learning

No data will be modified during this phase.

In [1]:
from pathlib import Path

import pandas as pd

In [2]:
# Define the project paths

current_directory = Path.cwd()

project_root = (
    current_directory.parent
    if current_directory.name.lower() == "notebooks"
    else current_directory
)

raw_data_path = (
    project_root
    / "data"
    / "raw"
    / "Walmart.csv"
)

print("Current directory:", current_directory)
print("Project root:", project_root)
print("Dataset path:", raw_data_path)

Current directory: c:\Users\Udoch\walmart-sales-regression\notebooks
Project root: c:\Users\Udoch\walmart-sales-regression
Dataset path: c:\Users\Udoch\walmart-sales-regression\data\raw\Walmart.csv


In [3]:
# Confirm the dataset exists

if not raw_data_path.exists():
    raise FileNotFoundError(
        f"The dataset was not found at: {raw_data_path}"
    )

print("Dataset found successfully.")

Dataset found successfully.


In [4]:
# Load a fresh copy of the dataset

df = pd.read_csv(raw_data_path)

print("Dataset loaded successfully.")

Dataset loaded successfully.


## Preview the First Records

The first five rows help confirm that the dataset loaded correctly and provide an initial view of the available columns and values.

In [5]:
df.head()

,Store,Date,Weekly_Sales,Holiday_Flag,Temperature,Fuel_Price,CPI,Unemployment
0,1,05-02-2010,1643690.90,0,42.31,2.572,211.096358,8.106
1,1,12-02-2010,1641957.44,1,38.51,2.548,211.242170,8.106
2,1,19-02-2010,1611968.17,0,39.93,2.514,211.289143,8.106
3,1,26-02-2010,1409727.59,0,46.63,2.561,211.319643,8.106
4,1,05-03-2010,1554806.68,0,46.50,2.625,211.350143,8.106


In [6]:
# Preview the Last Records

# The last five rows help confirm that the dataset was loaded completely and allow the final observations to be inspected.

df.tail()

,Store,Date,Weekly_Sales,Holiday_Flag,Temperature,Fuel_Price,CPI,Unemployment
6430,45,28-09-2012,713173.95,0,64.88,3.997,192.013558,8.684
6431,45,05-10-2012,733455.07,0,64.89,3.985,192.170412,8.667
6432,45,12-10-2012,734464.36,0,54.47,4.000,192.327265,8.667
6433,45,19-10-2012,718125.53,0,56.47,3.969,192.330854,8.667
6434,45,26-10-2012,760281.43,0,58.85,3.882,192.308899,8.667


## Dataset Dimensions

The dataset shape reports the number of rows and columns.

- Rows represent observations.
- Columns represent variables or features.

In [7]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 6435
Number of columns: 8


### Interpretation

The dataset contains 6,435 observations and 8 columns.

Each observation represents the weekly sales record for a particular Walmart store on a particular date.

In [8]:
# Inspect the column names
#Column Names

# Inspecting the column names helps confirm the available variables and identify the target column.

df.columns.tolist()

['Store',
 'Date',
 'Weekly_Sales',
 'Holiday_Flag',
 'Temperature',
 'Fuel_Price',
 'CPI',
 'Unemployment']

# Create a data dictionary

## Data Dictionary

| Column | Description | Expected Type |
|---|---|---|
| Store | Unique Walmart store identifier | Categorical identifier |
| Date | Weekly sales observation date | Datetime |
| Weekly_Sales | Weekly sales amount | Numerical continuous |
| Holiday_Flag | Indicates whether the week includes a holiday | Binary categorical |
| Temperature | Average temperature for the week | Numerical continuous |
| Fuel_Price | Regional fuel price | Numerical continuous |
| CPI | Consumer Price Index | Numerical continuous |
| Unemployment | Unemployment rate | Numerical continuous |

# Inspect data types

## Data Types

Data types help determine how each column is currently stored.

Incorrect types may need correction during the cleaning phase.

In [9]:
df.dtypes

Store             int64
Date                str
Weekly_Sales    float64
Holiday_Flag      int64
Temperature     float64
Fuel_Price      float64
CPI             float64
Unemployment    float64
dtype: object

### Data-Type Interpretation

- `Store` is stored as an integer, but it represents an identifier rather than a measured quantity.
- `Date` is stored as an object and should later be converted to a datetime type.
- `Weekly_Sales` is numerical and continuous.
- `Holiday_Flag` is stored as an integer but represents a binary category.
- Temperature, fuel price, CPI, and unemployment are numerical variables.

In [10]:
# Inspect full dataset information

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         6435 non-null   int64  
 1   Date          6435 non-null   str    
 2   Weekly_Sales  6435 non-null   float64
 3   Holiday_Flag  6435 non-null   int64  
 4   Temperature   6435 non-null   float64
 5   Fuel_Price    6435 non-null   float64
 6   CPI           6435 non-null   float64
 7   Unemployment  6435 non-null   float64
dtypes: float64(5), int64(2), str(1)
memory usage: 465.2 KB


### Dataset Information Interpretation

The `info()` output provides a quick summary of the dataset structure.

The non-null counts can be compared with the total number of rows to identify possible missing values.

The `Date` column currently requires conversion from object to datetime during data cleaning.

# Check missing values

## Missing-Value Assessment

Missing values can reduce model quality and may require removal, replacement, or further investigation.

In [11]:
# Check missing values

missing_values = df.isna().sum()

missing_values

Store           0
Date            0
Weekly_Sales    0
Holiday_Flag    0
Temperature     0
Fuel_Price      0
CPI             0
Unemployment    0
dtype: int64

In [12]:
missing_percentage = (
    df.isna().mean() * 100
).round(2)

missing_summary = pd.DataFrame(
    {
        "Missing_Count": missing_values,
        "Missing_Percentage": missing_percentage,
    }
)

missing_summary

,Missing_Count,Missing_Percentage
Store,0,0.0
Date,0,0.0
Weekly_Sales,0,0.0
Holiday_Flag,0,0.0
Temperature,0,0.0
Fuel_Price,0,0.0
CPI,0,0.0
Unemployment,0,0.0


### Missing-Value Interpretation

The missing-value assessment shows that all columns contain:

- **0 missing values**
- **0.0% missing data**

This means the dataset is complete and no missing-value treatment is currently required.

Therefore, techniques such as dropping rows, filling values with the mean, median, or mode, or using an imputer are not needed at this stage.

However, missing-value checks will still remain part of the preprocessing pipeline so that the final model can handle unexpected missing inputs more safely during deployment.

# Counting the unique values

## Unique Values

The number of unique values helps distinguish identifiers, categories, and continuous variables.

In [13]:
# coding 

unique_values = df.nunique()

unique_values

Store             45
Date             143
Weekly_Sales    6435
Holiday_Flag       2
Temperature     3528
Fuel_Price       892
CPI             2145
Unemployment     349
dtype: int64

In [ ]:
# Or ocde with this

unique_summary = pd.DataFrame(
    {
        "Column": df.columns,
        "Unique_Values": [
            df[column].nunique()
            for column in df.columns
        ],
    }
)

unique_summary

,Column,Unique_Values
0,Store,45
1,Date,143
2,Weekly_Sales,6435
3,Holiday_Flag,2
4,Temperature,3528
5,Fuel_Price,892
6,CPI,2145
7,Unemployment,349


### Unique-Value Interpretation

The unique-value analysis provides a better understanding of the structure of each feature in the dataset.

#### Observations

- **Store** contains **45 unique values**, indicating that the dataset includes sales records from 45 different Walmart stores.

- **Date** contains **143 unique values**, showing that sales were recorded over 143 different dates throughout the study period.

- **Weekly_Sales** contains **6,435 unique values**, meaning almost every sales record has a different sales amount. This is expected because it is the target variable for the regression model.

- **Holiday_Flag** contains only **2 unique values**, representing whether a week is a holiday or not (0 = Non-Holiday, 1 = Holiday). This is a binary categorical feature.

- **Temperature** contains **3,528 unique values**, indicating that temperature is a continuous numerical variable with many different recorded values.

- **Fuel_Price** contains **892 unique values**, showing that fuel prices changed over time and should be treated as a continuous numerical feature.

- **CPI** contains **2,145 unique values**, indicating continuous economic measurements across different dates and stores.

- **Unemployment** contains **349 unique values**, representing changing unemployment rates over time and across locations.

#### Conclusion

The dataset contains a good combination of:

- Identifier variables (Store)
- Time-related variables (Date)
- Binary categorical variables (Holiday_Flag)
- Continuous numerical variables (Temperature, Fuel_Price, CPI, Unemployment)
- A continuous target variable (Weekly_Sales)

This confirms that the dataset has the expected structure for a **supervised regression problem**, where the objective is to predict **Weekly_Sales** using the remaining features.

In [15]:
df.dtypes

Store             int64
Date                str
Weekly_Sales    float64
Holiday_Flag      int64
Temperature     float64
Fuel_Price      float64
CPI             float64
Unemployment    float64
dtype: object

In [16]:
# Inspecting the Store values

store_values = sorted(
    df["Store"].unique()
)

print("Number of stores:", len(store_values))
print("Store identifiers:", store_values)

Number of stores: 45
Store identifiers: [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12), np.int64(13), np.int64(14), np.int64(15), np.int64(16), np.int64(17), np.int64(18), np.int64(19), np.int64(20), np.int64(21), np.int64(22), np.int64(23), np.int64(24), np.int64(25), np.int64(26), np.int64(27), np.int64(28), np.int64(29), np.int64(30), np.int64(31), np.int64(32), np.int64(33), np.int64(34), np.int64(35), np.int64(36), np.int64(37), np.int64(38), np.int64(39), np.int64(40), np.int64(41), np.int64(42), np.int64(43), np.int64(44), np.int64(45)]


### Store Interpretation

The dataset contains 45 Walmart stores.

Store numbers are identifiers, not numerical quantities. Therefore, `Store` will later be treated as a categorical feature.

In [17]:
# Inspect Holiday Flag values

df["Holiday_Flag"].value_counts()

Holiday_Flag
0    5985
1     450
Name: count, dtype: int64

In [18]:
df["Holiday_Flag"].unique()

array([0, 1])

### Holiday Flag Interpretation

`Holiday_Flag` contains two values:

- `0`: Non-holiday week
- `1`: Holiday week

This confirms that the feature is binary.

# Generating summary statistics

## Numerical Summary Statistics

Summary statistics describe the centre, spread, minimum, maximum, and distribution of numerical variables.

In [20]:
# coding 

df.describe().T

,count,mean,std,min,25%,50%,75%,max
Store,6435.0,2.300000e+01,12.988182,1.000,12.000,23.000000,3.400000e+01,4.500000e+01
Weekly_Sales,6435.0,1.046965e+06,564366.622054,209986.250,553350.105,960746.040000,1.420159e+06,3.818686e+06
Holiday_Flag,6435.0,6.993007e-02,0.255049,0.000,0.000,0.000000,0.000000e+00,1.000000e+00
Temperature,6435.0,6.066378e+01,18.444933,-2.060,47.460,62.670000,7.494000e+01,1.001400e+02
Fuel_Price,6435.0,3.358607e+00,0.459020,2.472,2.933,3.445000,3.735000e+00,4.468000e+00
CPI,6435.0,1.715784e+02,39.356712,126.064,131.735,182.616521,2.127433e+02,2.272328e+02
Unemployment,6435.0,7.999151e+00,1.875885,3.879,6.891,7.874000,8.622000e+00,1.431300e+01


### Summary Statistics Interpretation

The summary statistics provide an overview of the distribution, central tendency, and variability of the numerical features in the Walmart sales dataset.

#### Observations

- All numerical columns contain **6,435 observations**, confirming that there are no missing values in the dataset.

- The dataset contains **45 unique Walmart stores**, with store IDs ranging from **1 to 45**.

- **Weekly_Sales** range from approximately **£20,969** to **£381,869**, indicating substantial variation in weekly sales across stores and dates.

- The median (**50th percentile**) weekly sales is approximately **£96,071**, meaning that half of the weekly sales values are below this amount and half are above it.

- The large standard deviation (**£56,498**) for Weekly_Sales suggests considerable variability in sales performance between different stores and weeks.

- **Holiday_Flag** has values ranging from **0 to 1**, confirming that it is a binary feature indicating whether a week is a holiday or not.

- **Temperature** ranges from approximately **-2.06°C to 47.4°C**, showing that the dataset covers a wide range of weather conditions.

- **Fuel_Price** varies between **2.47** and **4.47**, indicating noticeable changes in fuel prices over the observation period.

- **CPI (Consumer Price Index)** ranges from approximately **126.06 to 227.23**, reflecting changes in economic conditions throughout the dataset.

- **Unemployment** ranges from approximately **3.88% to 14.31%**, suggesting that unemployment levels varied considerably across locations and time.

#### Conclusion

The summary statistics indicate that the dataset contains realistic business data with meaningful variation across stores, time periods, and economic conditions. No abnormal counts or missing observations are present.

The wide range observed in **Weekly_Sales** suggests that some stores or weeks achieved significantly higher sales than others. During the Exploratory Data Analysis (EDA) phase, visualizations such as histograms and boxplots will be used to investigate the distribution of sales and identify any potential outliers before model training.

## Target Variable

The target variable is the value the machine learning model will predict.

In [21]:
# Inspecting the target variable

target_column = "Weekly_Sales"

df[target_column].describe()

count    6.435000e+03
mean     1.046965e+06
std      5.643666e+05
min      2.099862e+05
25%      5.533501e+05
50%      9.607460e+05
75%      1.420159e+06
max      3.818686e+06
Name: Weekly_Sales, dtype: float64

### Target Variable Interpretation

The target variable for this project is **`Weekly_Sales`**, which represents the total weekly sales generated by each Walmart store.

#### Observations

- The dataset contains **6,435 sales records**, meaning there are no missing values in the target variable.

- The **average weekly sales** is approximately **£1,046,965**, indicating the typical weekly revenue across all stores.

- The **minimum weekly sales** is approximately **£209,986**, while the **maximum weekly sales** reaches approximately **£3,818,686**. This large difference suggests that sales vary considerably between stores and different weeks.

- The **median (50th percentile)** weekly sales is approximately **£960,746**, which is slightly lower than the mean. This indicates that some very high sales values increase the average.

- The **standard deviation** is approximately **£564,367**, showing that weekly sales are highly variable throughout the dataset.

- The first quartile (**25%**) is approximately **£553,350**, while the third quartile (**75%**) is approximately **£1,420,159**. This means that half of all weekly sales fall between these two values.

#### Conclusion

The target variable is **numeric and continuous**, confirming that this is a **supervised machine learning regression problem**.

The large range and high variability in weekly sales suggest that factors such as store location, holidays, temperature, fuel prices, CPI, and unemployment may significantly influence sales performance.

During the Exploratory Data Analysis (EDA) phase, histograms and boxplots will be used to examine the distribution of `Weekly_Sales` and identify any potential outliers before building predictive models.

In [22]:
# Inspect the date values

date_preview = pd.to_datetime(
    df["Date"],
    format="%d-%m-%Y",
    errors="coerce",
)

print("Earliest date:", date_preview.min())
print("Latest date:", date_preview.max())
print("Invalid dates:", date_preview.isna().sum())

Earliest date: 2010-02-05 00:00:00
Latest date: 2012-10-26 00:00:00
Invalid dates: 0


### Date-Range Interpretation

The date preview shows that the dataset covers the period from **5 February 2010** to **26 October 2012**.

A total of **0 invalid dates** were found, which means every value in the `Date` column was successfully interpreted using the expected `day-month-year` format.

This confirms that the date values are consistent and suitable for later conversion to a proper datetime data type.

Because the observations are time-based, the dataset should later be sorted by date and split chronologically for model training and evaluation. This will help prevent future information from leaking into the training data.

In [23]:
# Building a complete quality-summary table

quality_summary = pd.DataFrame(
    {
        "Data_Type": df.dtypes.astype(str),
        "Missing_Count": df.isna().sum(),
        "Missing_Percentage": (
            df.isna().mean() * 100
        ).round(2),
        "Unique_Values": df.nunique(),
    }
)

quality_summary

,Data_Type,Missing_Count,Missing_Percentage,Unique_Values
Store,int64,0,0.0,45
Date,str,0,0.0,143
Weekly_Sales,float64,0,0.0,6435
Holiday_Flag,int64,0,0.0,2
Temperature,float64,0,0.0,3528
Fuel_Price,float64,0,0.0,892
CPI,float64,0,0.0,2145
Unemployment,float64,0,0.0,349


# Separating expected feature types

## Initial Feature Classification

Based on the dataset structure, the variables can be classified as follows:

### Target Variable

- `Weekly_Sales`

### Categorical or Identifier Features

- `Store`
- `Holiday_Flag`

### Date Feature

- `Date`

### Numerical Features

- `Temperature`
- `Fuel_Price`
- `CPI`
- `Unemployment`

The date column will later be used to create calendar features such as year, month, and week of year.

## Key Findings

- The dataset contains historical weekly sales observations for 45 Walmart stores.
- Each row represents one store on one weekly date.
- `Weekly_Sales` is the target variable.
- The target is numerical and continuous, confirming a regression problem.
- `Store` is an identifier and should later be treated as categorical.
- `Holiday_Flag` is a binary variable.
- `Date` currently requires conversion to datetime.
- Missing values and duplicates were assessed.
- The data contains a time dimension, so chronological model evaluation will be important.
- Potential outliers should be investigated before any removal decision is made.

## Phase Conclusion

The Walmart dataset has been successfully examined without modifying the original data.

The dataset structure, dimensions, columns, data types, missing values, duplicates, unique values, summary statistics, date range, and target variable were reviewed.

The next phase is **Data Cleaning and Validation**, where the project will:

- Standardise column names
- Convert the date column
- Validate required columns
- Check invalid values
- Handle duplicates where necessary
- Investigate potential outliers
- Save a cleaned dataset in `data/processed/`